src/components/verification-grid/verification-grid.ts

```diff
@@ -217,8 +217,20 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
 
   protected async updated(change: PropertyValueMap<this>): Promise<void> {
     const renderInvalidationKeys: (keyof this)[] = ["gridSize"];
-    const sourceInvalidationKeys: (keyof this)[] = ["getPage"];
     const tileInvalidationKeys: (keyof this)[] = ["selectionBehavior"];
+    // gridSize is a part of source invalidation because if the grid size
+    // increases, there will be verification grid tiles without any source
+    // additionally, if the grid size is decreased, we want the "currentPage"
+    // of sources to update / remove un-needed items
+    const sourceInvalidationKeys: (keyof this)[] = ["getPage", "gridSize"];
+
+    // a full render invalidation causes all the grid tiles elements to be
+    // destroyed and re-created. This should always be a last resort
+    // we do this first so that if there is a source or tile invalidation
+    // the changes are applied to the newly created grid tile elements
```

Comment:
since when do we destroy tiles?

Reply:
We currently destroy all the verification grid tiles during a render invalidation

I have an inline todo that we should implement a partial re-rendering + diffing algorithm instead of destroying all the tiles during a render invalidation

![image](https://github.com/user-attachments/assets/aed8ef43-1afa-46f6-af2c-ab13ee985a6a)

Reply:
Update: I implemented the proposed Diff algorithm (it was quite simple to implement). We no longer destroy grid tiles

![image](https://github.com/user-attachments/assets/d45d110d-7cfe-439b-baf2-ffeea071f820)

Reply:
i am wondering why the lit `repeat` directive is not at the center of this strategy

but yes, probably a separate PR

Reply:
Related issue: #149 (linked in "TODO" comment as well)